File: tests/test_lexer.py
import pytest

from ..harpy.lexer import Lexer
from ..harpy.token import Token
from ..harpy.token_type import TokenType


class TestLexer:
    def test__next__(self):
        obs = self._get_obs(source="from + offset(time)")
        expected = [
            Token(type=TokenType.NAME, text="from", line=1, position=4),
            Token(type=TokenType.PLUS, text="+", line=1, position=6),
            Token(type=TokenType.NAME, text="offset", line=1, position=13),
            Token(type=TokenType.LEFT_PAREN, text="(", line=1, position=14),
            Token(type=TokenType.NAME, text="time", line=1, position=18),
            Token(type=TokenType.RIGHT_PAREN, text=")", line=1, position=19),
        ]

        assert str(obs) == str(expected)

    def test_preprocessor_directive(self):
        obs = self._get_obs(source="#define slBool .f.")
        expected = [
            Token(
                type=TokenType.DEFINE_DIRECTIVE,
                text="#define slBool .f.",
                line=1,
                position=18,
            )
        ]

        assert str(obs) == str(expected)

        obs = self._get_obs(source="#ifdef SOMETHING")
        expected = [
            Token(
                type=TokenType.IFDEF_DIRECTIVE,
                text="#ifdef SOMETHING",
                line=1,
                position=16,
            )
        ]

        assert str(obs) == str(expected)

        obs = self._get_obs(source="#pragma -ko+")
        expected = [
            Token(
                type=TokenType.PRAGMA_DIRECTIVE, text="#pragma -ko+", line=1, position=12
            )
        ]

        assert str(obs) == str(expected)

    def test_bool_literal(self):
        obs = self._get_obs(source=".t.")
        expected = [Token(type=TokenType.BOOL_LITERAL, text=".t.", line=1, position=3)]

        assert str(obs) == str(expected)

        obs = self._get_obs(source=".F.")
        expected = [Token(type=TokenType.BOOL_LITERAL, text=".F.", line=1, position=3)]

        assert str(obs) == str(expected)

    def test_num_literal(self):
        obs = self._get_obs(source="0")
        expected = [Token(type=TokenType.NUM_LITERAL, text="0", line=1, position=1)]

        assert str(obs) == str(expected)

        obs = self._get_obs(source="123")
        expected = [Token(type=TokenType.NUM_LITERAL, text="123", line=1, position=3)]

        assert str(obs) == str(expected)

        obs = self._get_obs(source="50000")
        expected = [Token(type=TokenType.NUM_LITERAL, text="50000", line=1, position=5)]

        assert str(obs) == str(expected)

        obs = self._get_obs(source="12000.123")
        expected = [
            Token(type=TokenType.NUM_LITERAL, text="12000.123", line=1, position=9)
        ]

        assert str(obs) == str(expected)

        obs = self._get_obs(source="0xABAB")
        expected = [
            Token(type=TokenType.NUM_LITERAL, text="0xABAB", line=1, position=6)
        ]

        assert str(obs) == str(expected)

        obs = self._get_obs(source=".12")
        expected = [Token(type=TokenType.NUM_LITERAL, text=".12", line=1, position=3)]

        assert str(obs) == str(expected)

        with pytest.raises(SyntaxError, match="Unterminated hexadecimal literal '1x'"):
            _ = self._get_obs(source="1x001")

        with pytest.raises(SyntaxError, match="Invalid numeric literal '123a'"):
            _ = self._get_obs(source="123a")

        with pytest.raises(
            SyntaxError, match="Second decimal point found in literal '1.1.'."
        ):
            _ = self._get_obs(source="1.1.1")

    def test_str_literal(self):
        obs = self._get_obs(source="'This is a string'")
        expected = [
            Token(
                type=TokenType.STR_LITERAL,
                text="'This is a string'",
                line=1,
                position=18,
            )
        ]

        assert str(obs) == str(expected)

        obs = self._get_obs(source='"This is also a string"')
        expected = [
            Token(TokenType.STR_LITERAL, '"This is also a string"', line=1, position=23)
        ]

        assert str(obs) == str(expected)

        obs = self._get_obs(source="[Also a string]")
        expected = [
            Token(
                type=TokenType.STR_LITERAL, text="[Also a string]", line=1, position=15
            )
        ]

        assert str(obs) == str(expected)

        obs = self._get_obs(source="['Actually a hash key']")
        expected = [
            Token(type=TokenType.LEFT_BRACKET, text="[", line=1, position=1),
            Token(
                type=TokenType.STR_LITERAL,
                text="'Actually a hash key'",
                line=1,
                position=22,
            ),
            Token(type=TokenType.RIGHT_BRACKET, text="]", line=1, position=23),
        ]

        assert str(obs) == str(expected)

        obs = self._get_obs(source="function a(b)\n    local i := 1\nreturn b[i]")
        expected = [
            Token(type=TokenType.FUNCTION, text="function", line=1, position=8),
            Token(type=TokenType.NAME, text="a", line=1, position=10),
            Token(type=TokenType.LEFT_PAREN, text="(", line=1, position=11),
            Token(type=TokenType.NAME, text="b", line=1, position=12),
            Token(type=TokenType.RIGHT_PAREN, text=")", line=1, position=13),
            Token(type=TokenType.LOCAL, text="local", line=2, position=9),
            Token(type=TokenType.NAME, text="i", line=2, position=11),
            Token(type=TokenType.ASSIGN, text=":=", line=2, position=14),
            Token(type=TokenType.NUM_LITERAL, text="1", line=2, position=16),
            Token(type=TokenType.RETURN, text="return", line=3, position=6),
            Token(type=TokenType.NAME, text="b", line=3, position=8),
            Token(type=TokenType.LEFT_BRACKET, text="[", line=3, position=9),
            Token(type=TokenType.NAME, text="i", line=3, position=10),
            Token(type=TokenType.RIGHT_BRACKET, text="]", line=3, position=11),
        ]

        assert str(obs) == str(expected)

        obs = self._get_obs(source='""')
        expected = [Token(TokenType.STR_LITERAL, '""', line=1, position=2)]

        assert str(obs) == str(expected)

        obs = self._get_obs(source="''")
        expected = [Token(type=TokenType.STR_LITERAL, text="''", line=1, position=2)]

        assert str(obs) == str(expected)

        with pytest.raises(SyntaxError):
            _ = self._get_obs(source="'This string does not finish")

        with pytest.raises(SyntaxError):
            _ = self._get_obs(source="'")

    def test_assign_vs_comma(self):
        obs = self._get_obs(source="a := b")
        expected = [
            Token(type=TokenType.NAME, text="a", line=1, position=1),
            Token(type=TokenType.ASSIGN, text=":=", line=1, position=4),
            Token(type=TokenType.NAME, text="b", line=1, position=6),
        ]

        assert str(obs) == str(expected)

        obs = self._get_obs(source="a:b")
        expected = [
            Token(type=TokenType.NAME, text="a", line=1, position=1),
            Token(type=TokenType.COLON, text=":", line=1, position=2),
            Token(type=TokenType.NAME, text="b", line=1, position=3),
        ]

        assert str(obs) == str(expected)

    def test_logical_operators(self):
        obs = self._get_obs(source="a .and. b")
        expected = [
            Token(type=TokenType.NAME, text="a", line=1, position=1),
            Token(type=TokenType.AND, text=".and.", line=1, position=7),
            Token(type=TokenType.NAME, text="b", line=1, position=9),
        ]

        assert str(obs) == str(expected)

        obs = self._get_obs(source="a .and. !b")
        expected = [
            Token(type=TokenType.NAME, text="a", line=1, position=1),
            Token(type=TokenType.AND, text=".and.", line=1, position=7),
            Token(type=TokenType.NOT, text="!", line=1, position=9),
            Token(type=TokenType.NAME, text="b", line=1, position=10),
        ]

        assert str(obs) == str(expected)

        obs = self._get_obs(source="!a .and. !b")
        expected = [
            Token(type=TokenType.NOT, text="!", line=1, position=1),
            Token(type=TokenType.NAME, text="a", line=1, position=2),
            Token(type=TokenType.AND, text=".and.", line=1, position=8),
            Token(type=TokenType.NOT, text="!", line=1, position=10),
            Token(type=TokenType.NAME, text="b", line=1, position=11),
        ]

        assert str(obs) == str(expected)

        obs = self._get_obs(source="!a .and. !b .or. !c")
        expected = [
            Token(type=TokenType.NOT, text="!", line=1, position=1),
            Token(type=TokenType.NAME, text="a", line=1, position=2),
            Token(type=TokenType.AND, text=".and.", line=1, position=8),
            Token(type=TokenType.NOT, text="!", line=1, position=10),
            Token(type=TokenType.NAME, text="b", line=1, position=11),
            Token(type=TokenType.OR, text=".or.", line=1, position=16),
            Token(type=TokenType.NOT, text="!", line=1, position=18),
            Token(type=TokenType.NAME, text="c", line=1, position=19),
        ]

        assert str(obs) == str(expected)

    def test_relations(self):
        obs = self._get_obs(source="a == b")
        expected = [
            Token(type=TokenType.NAME, text="a", line=1, position=1),
            Token(type=TokenType.EQ1, text="==", line=1, position=4),
            Token(type=TokenType.NAME, text="b", line=1, position=6),
        ]

        assert str(obs) == str(expected)

        obs = self._get_obs(source="a <= b")
        expected = [
            Token(type=TokenType.NAME, text="a", line=1, position=1),
            Token(type=TokenType.LE, text="<=", line=1, position=4),
            Token(type=TokenType.NAME, text="b", line=1, position=6),
        ]

        assert str(obs) == str(expected)

    def test_line_comment(self):
        obs = self._get_obs(source="// This is a line comment.")
        expected = [
            Token(
                type=TokenType.LINE_COMMENT,
                text="// This is a line comment.",
                line=1,
                position=26,
            )
        ]

        assert str(obs) == str(expected)

    def test_block_comment(self):
        obs = self._get_obs(source="/* This is a block comment.*/")
        expected = [
            Token(
                type=TokenType.BLOCK_COMMENT,
                text="/* This is a block comment.*/",
                line=1,
                position=29,
            )
        ]

        assert str(obs) == str(expected)

        obs = self._get_obs(
            source="""
/* This is also a
 * block
 * comment.
 */
 """
        )
        expected = [
            Token(
                TokenType.BLOCK_COMMENT,
                "/* This is also a\n * block\n * comment.\n */",
                line=2,
                position=0,
            )
        ]

        assert str(obs) == str(expected)

        with pytest.raises(SyntaxError):
            self._get_obs(source="/* This is an unfinished block comment.*")

        with pytest.raises(SyntaxError):
            self._get_obs(source="/* This one is even worse./*")

    def test_function_def(self):
        obs = self._get_obs(source="FUNCTION a()\nRETURN b")
        expected = [
            Token(type=TokenType.FUNCTION, text="FUNCTION", line=1, position=8),
            Token(type=TokenType.NAME, text="a", line=1, position=10),
            Token(type=TokenType.LEFT_PAREN, text="(", line=1, position=11),
            Token(type=TokenType.RIGHT_PAREN, text=")", line=1, position=12),
            Token(type=TokenType.RETURN, text="RETURN", line=2, position=6),
            Token(type=TokenType.NAME, text="b", line=2, position=8),
        ]

        assert str(obs) == str(expected)

        obs = self._get_obs(source="function a(b, c, d)\n\nreturn e")
        expected = [
            Token(type=TokenType.FUNCTION, text="function", line=1, position=8),
            Token(type=TokenType.NAME, text="a", line=1, position=10),
            Token(type=TokenType.LEFT_PAREN, text="(", line=1, position=11),
            Token(type=TokenType.NAME, text="b", line=1, position=12),
            Token(type=TokenType.COMMA, text=",", line=1, position=13),
            Token(type=TokenType.NAME, text="c", line=1, position=15),
            Token(type=TokenType.COMMA, text=",", line=1, position=16),
            Token(type=TokenType.NAME, text="d", line=1, position=18),
            Token(type=TokenType.RIGHT_PAREN, text=")", line=1, position=19),
            Token(type=TokenType.RETURN, text="return", line=3, position=6),
            Token(type=TokenType.NAME, text="e", line=3, position=8),
        ]

        assert str(obs) == str(expected)

    def test_procedure_def(self):
        obs = self._get_obs(source="procedure a()\nRETURN")
        expected = [
            Token(type=TokenType.PROCEDURE, text="procedure", line=1, position=9),
            Token(type=TokenType.NAME, text="a", line=1, position=11),
            Token(type=TokenType.LEFT_PAREN, text="(", line=1, position=12),
            Token(type=TokenType.RIGHT_PAREN, text=")", line=1, position=13),
            Token(type=TokenType.RETURN, text="RETURN", line=2, position=6),
        ]

        assert str(obs) == str(expected)

        obs = self._get_obs(source="procedure a(b, c, d)\n\nreturn")
        expected = [
            Token(type=TokenType.PROCEDURE, text="procedure", line=1, position=9),
            Token(type=TokenType.NAME, text="a", line=1, position=11),
            Token(type=TokenType.LEFT_PAREN, text="(", line=1, position=12),
            Token(type=TokenType.NAME, text="b", line=1, position=13),
            Token(type=TokenType.COMMA, text=",", line=1, position=14),
            Token(type=TokenType.NAME, text="c", line=1, position=16),
            Token(type=TokenType.COMMA, text=",", line=1, position=17),
            Token(type=TokenType.NAME, text="d", line=1, position=19),
            Token(type=TokenType.RIGHT_PAREN, text=")", line=1, position=20),
            Token(type=TokenType.RETURN, text="return", line=3, position=6),
        ]

        assert str(obs) == str(expected)

    def test_if_else(self):
        obs = self._get_obs(
            source="if a <= b\n    b()\nelseif c\n    d()\nelse\n    e()\nendif"
        )
        expected = [
            Token(type=TokenType.IF, text="if", line=1, position=2),
            Token(type=TokenType.NAME, text="a", line=1, position=4),
            Token(type=TokenType.LE, text="<=", line=1, position=7),
            Token(type=TokenType.NAME, text="b", line=1, position=9),
            Token(type=TokenType.NAME, text="b", line=2, position=5),
            Token(type=TokenType.LEFT_PAREN, text="(", line=2, position=6),
            Token(type=TokenType.RIGHT_PAREN, text=")", line=2, position=7),
            Token(type=TokenType.ELSEIF, text="elseif", line=3, position=6),
            Token(type=TokenType.NAME, text="c", line=3, position=8),
            Token(type=TokenType.NAME, text="d", line=4, position=5),
            Token(type=TokenType.LEFT_PAREN, text="(", line=4, position=6),
            Token(type=TokenType.RIGHT_PAREN, text=")", line=4, position=7),
            Token(type=TokenType.ELSE, text="else", line=5, position=4),
            Token(type=TokenType.NAME, text="e", line=6, position=5),
            Token(type=TokenType.LEFT_PAREN, text="(", line=6, position=6),
            Token(type=TokenType.RIGHT_PAREN, text=")", line=6, position=7),
            Token(type=TokenType.ENDIF, text="endif", line=7, position=5),
        ]

        assert str(obs) == str(expected)

    def _get_obs(self, source: str) -> list[Token]:
        lexer = Lexer(text=source)
        obs = []
        for token in lexer:
            if token.type == TokenType.EOF:
                break
            else:
                obs.append(token)

        return obs

File: HarpyTests/TestLexer.cs
using System.Data;

using Harpy.Lexer;

namespace HarpyTests
{
    [TestClass]
    public sealed class TestLexer
    {
        private static Lexer? lexer;

        [TestMethod]
        public void TestGetTokens()
        {
            var obs = GetObservedTokens("from + offset(time)");
            var expected = new List<Token> {
                new(SyntaxType.NAME, "from", 1, 4),
                new(SyntaxType.PLUS, "+", 1, 6),
                new(SyntaxType.NAME, "offset", 1, 13),
                new(SyntaxType.LEFT_PAREN, "(", 1, 14),
                new(SyntaxType.NAME, "time", 1, 18),
                new(SyntaxType.RIGHT_PAREN, ")", 1, 19)
            };

            AssertEqual(obs, expected);
        }

        [TestMethod]
        public void TestPreprocessorDirective()
        {
            var obs = GetObservedTokens("#define slBool .f.");
            var expected = new List<Token> {
                new(
                    SyntaxType.DEFINE_DIRECTIVE,
                    "#define slBool .f.",
                    1,
                    18
                )
            };

            AssertEqual(obs, expected);

            obs = GetObservedTokens("#ifdef SOMETHING");
            expected = [
                    new(
                    SyntaxType.IFDEF_DIRECTIVE,
                    "#ifdef SOMETHING",
                    1,
                    16
                )
                ];

            AssertEqual(obs, expected);

            obs = GetObservedTokens("#pragma -ko+");
            expected = [
                    new(
                    SyntaxType.PRAGMA_DIRECTIVE,
                    "#pragma -ko+",
                    1,
                    12
                )
                ];

            AssertEqual(obs, expected);
        }

        [TestMethod]
        public void TestBooleanLiteral()
        {
            var obs = GetObservedTokens(".t.");
            var expected = new List<Token> { new(SyntaxType.BOOL_LITERAL, ".t.", 1, 3) };

            AssertEqual(obs, expected);

            obs = GetObservedTokens(".F.");
            expected = [new(SyntaxType.BOOL_LITERAL, ".F.", 1, 3)];

            AssertEqual(obs, expected);
        }

        [TestMethod]
        public void TestNumericLiteral()
        {
            var obs = GetObservedTokens("0");
            var expected = new List<Token> { new(SyntaxType.NUM_LITERAL, "0", 1, 1) };

            AssertEqual(obs, expected);

            obs = GetObservedTokens("123");
            expected = [new(SyntaxType.NUM_LITERAL, "123", 1, 3)];

            AssertEqual(obs, expected);

            obs = GetObservedTokens("50000");
            expected = [new(SyntaxType.NUM_LITERAL, "50000", 1, 5)];

            AssertEqual(obs, expected);

            obs = GetObservedTokens("12000.123");
            expected = [
                    new(SyntaxType.NUM_LITERAL, "12000.123", 1, 9)
                ];

            AssertEqual(obs, expected);

            obs = GetObservedTokens("0xABAB");
            expected = [
                    new(SyntaxType.NUM_LITERAL, "0xABAB", 1, 6)
                ];

            AssertEqual(obs, expected);

            obs = GetObservedTokens(".12");
            expected = [new(SyntaxType.NUM_LITERAL, ".12", 1, 3)];

            AssertEqual(obs, expected);

            Assert.ThrowsException<SyntaxErrorException>(() => GetObservedTokens("1x001"));

            Assert.ThrowsException<SyntaxErrorException>(() => GetObservedTokens("123a"));

            Assert.ThrowsException<SyntaxErrorException>(() => GetObservedTokens("1.1.1"));
        }

        [TestMethod]
        public void TestStringLiteral()
        {
            var obs = GetObservedTokens("'This is a string'");
            var expected = new List<Token> {
                new(
                    SyntaxType.STR_LITERAL,
                    "'This is a string'",
                    1,
                    18
                )
            };

            AssertEqual(obs, expected);

            obs = GetObservedTokens("\"This is also a string\"");
            expected = [
                    new(SyntaxType.STR_LITERAL, "\"This is also a string\"", 1, 23)
                ];

            AssertEqual(obs, expected);

            obs = GetObservedTokens("[Also a string]");
            expected = [
                    new(
                    SyntaxType.STR_LITERAL, "[Also a string]", 1, 15
                )
                ];

            AssertEqual(obs, expected);

            obs = GetObservedTokens("['Actually a hash key']");
            expected = [
                    new(SyntaxType.LEFT_BRACKET, "[", 1, 1),
                new(
                    SyntaxType.STR_LITERAL,
                    "'Actually a hash key'",
                    1,
                    22
                ),
                new(SyntaxType.RIGHT_BRACKET, "]", 1, 23),
            ];

            AssertEqual(obs, expected);

            obs = GetObservedTokens("function a(b)\n    local i := 1\nreturn b[i]");
            expected = [
                    new(SyntaxType.FUNCTION, "function", 1, 8),
                new(SyntaxType.NAME, "a", 1, 10),
                new(SyntaxType.LEFT_PAREN, "(", 1, 11),
                new(SyntaxType.NAME, "b", 1, 12),
                new(SyntaxType.RIGHT_PAREN, ")", 1, 13),
                new(SyntaxType.LOCAL, "local", 2, 9),
                new(SyntaxType.NAME, "i", 2, 11),
                new(SyntaxType.ASSIGN, ":=", 2, 14),
                new(SyntaxType.NUM_LITERAL, "1", 2, 16),
                new(SyntaxType.RETURN, "return", 3, 6),
                new(SyntaxType.NAME, "b", 3, 8),
                new(SyntaxType.LEFT_BRACKET, "[", 3, 9),
                new(SyntaxType.NAME, "i", 3, 10),
                new(SyntaxType.RIGHT_BRACKET, "]", 3, 11),
            ];

            AssertEqual(obs, expected);

            obs = GetObservedTokens("\"\"");
            expected = [new(SyntaxType.STR_LITERAL, "\"\"", 1, 2)];

            AssertEqual(obs, expected);

            obs = GetObservedTokens("''");
            expected = [new(SyntaxType.STR_LITERAL, "''", 1, 2)];

            AssertEqual(obs, expected);

            Assert.ThrowsException<SyntaxErrorException>(() => GetObservedTokens("'This string does not finish"));

            Assert.ThrowsException<SyntaxErrorException>(() => GetObservedTokens("'"));
        }

        [TestMethod]
        public void TestAssignVsComma()
        {
            var obs = GetObservedTokens("a := b");
            var expected = new List<Token> {
                new(SyntaxType.NAME, "a", 1, 1),
                new(SyntaxType.ASSIGN, ":=", 1, 4),
                new(SyntaxType.NAME, "b", 1, 6),
            };

            AssertEqual(obs, expected);

            obs = GetObservedTokens("a:b");
            expected = [
                    new(SyntaxType.NAME, "a", 1, 1),
                new(SyntaxType.COLON, ":", 1, 2),
                new(SyntaxType.NAME, "b", 1, 3),
            ];

            AssertEqual(obs, expected);
        }

        [TestMethod]
        public void TestLogicalOperators()
        {
            var obs = GetObservedTokens("a .and. b");
            var expected = new List<Token> {
                new(SyntaxType.NAME, "a", 1, 1),
                new(SyntaxType.AND, ".and.", 1, 7),
                new(SyntaxType.NAME, "b", 1, 9),
            };

            AssertEqual(obs, expected);

            obs = GetObservedTokens("a .and. !b");
            expected = [
                    new(SyntaxType.NAME, "a", 1, 1),
                new(SyntaxType.AND, ".and.", 1, 7),
                new(SyntaxType.NOT, "!", 1, 9),
                new(SyntaxType.NAME, "b", 1, 10),
            ];

            AssertEqual(obs, expected);

            obs = GetObservedTokens("!a .and. !b");
            expected = [
                    new(SyntaxType.NOT, "!", 1, 1),
                new(SyntaxType.NAME, "a", 1, 2),
                new(SyntaxType.AND, ".and.", 1, 8),
                new(SyntaxType.NOT, "!", 1, 10),
                new(SyntaxType.NAME, "b", 1, 11),
            ];

            AssertEqual(obs, expected);

            obs = GetObservedTokens("!a .and. !b .or. !c");
            expected = [
                    new(SyntaxType.NOT, "!", 1, 1),
                new(SyntaxType.NAME, "a", 1, 2),
                new(SyntaxType.AND, ".and.", 1, 8),
                new(SyntaxType.NOT, "!", 1, 10),
                new(SyntaxType.NAME, "b", 1, 11),
                new(SyntaxType.OR, ".or.", 1, 16),
                new(SyntaxType.NOT, "!", 1, 18),
                new(SyntaxType.NAME, "c", 1, 19),
            ];

            AssertEqual(obs, expected);
        }

        [TestMethod]
        public void TestRelations()
        {
            var obs = GetObservedTokens("a == b");
            var expected = new List<Token> {
                new(SyntaxType.NAME, "a", 1, 1),
                new(SyntaxType.EQ1, "==", 1, 4),
                new(SyntaxType.NAME, "b", 1, 6),
            };

            AssertEqual(obs, expected);

            obs = GetObservedTokens("a <= b");
            expected = [
                    new(SyntaxType.NAME, "a", 1, 1),
                new(SyntaxType.LE, "<=", 1, 4),
                new(SyntaxType.NAME, "b", 1, 6),
            ];

            AssertEqual(obs, expected);
        }

        [TestMethod]
        public void TestLineComment()
        {
            var obs = GetObservedTokens("// This is a line comment.");
            var expected = new List<Token> {
                new(
                    SyntaxType.LINE_COMMENT,
                    "// This is a line comment.",
                    1,
                    26
                )
            };

            AssertEqual(obs, expected);
        }

        [TestMethod]
        public void TestBlockComment()
        {
            var obs = GetObservedTokens("/* This is a block comment.*/");
            var expected = new List<Token> {
                new(
                    SyntaxType.BLOCK_COMMENT,
                    "/* This is a block comment.*/",
                    1,
                    29
                )
            };

            AssertEqual(obs, expected);

            obs = GetObservedTokens("\n/* This is also a\n * block\n * comment.\n */");
            expected = [
                    new(
                    SyntaxType.BLOCK_COMMENT,
                    "/* This is also a\n * block\n * comment.\n */",
                    2,
                    3
                )
                ];

            AssertEqual(obs, expected);

            Assert.ThrowsException<SyntaxErrorException>(() => GetObservedTokens("/* This is an unfinished block comment.*"));

            Assert.ThrowsException<SyntaxErrorException>(() => GetObservedTokens("/* This one is even worse./*"));
        }

        [TestMethod]
        public void TestFunctionDef()
        {
            var obs = GetObservedTokens("FUNCTION a()\nRETURN b");
            var expected = new List<Token> {
                new(SyntaxType.FUNCTION, "FUNCTION", 1, 8),
                new(SyntaxType.NAME, "a", 1, 10),
                new(SyntaxType.LEFT_PAREN, "(", 1, 11),
                new(SyntaxType.RIGHT_PAREN, ")", 1, 12),
                new(SyntaxType.RETURN, "RETURN", 2, 6),
                new(SyntaxType.NAME, "b", 2, 8),
            };

            AssertEqual(obs, expected);

            obs = GetObservedTokens("function a(b, c, d)\n\nreturn e");
            expected = [
                    new(SyntaxType.FUNCTION, "function", 1, 8),
                new(SyntaxType.NAME, "a", 1, 10),
                new(SyntaxType.LEFT_PAREN, "(", 1, 11),
                new(SyntaxType.NAME, "b", 1, 12),
                new(SyntaxType.COMMA, ",", 1, 13),
                new(SyntaxType.NAME, "c", 1, 15),
                new(SyntaxType.COMMA, ",", 1, 16),
                new(SyntaxType.NAME, "d", 1, 18),
                new(SyntaxType.RIGHT_PAREN, ")", 1, 19),
                new(SyntaxType.RETURN, "return", 3, 6),
                new(SyntaxType.NAME, "e", 3, 8),
            ];

            AssertEqual(obs, expected);
        }

        [TestMethod]
        public void TestProcedureDef()
        {
            var obs = GetObservedTokens("procedure a()\nRETURN");
            var expected = new List<Token> {
                new(SyntaxType.PROCEDURE, "procedure", 1, 9),
                new(SyntaxType.NAME, "a", 1, 11),
                new(SyntaxType.LEFT_PAREN, "(", 1, 12),
                new(SyntaxType.RIGHT_PAREN, ")", 1, 13),
                new(SyntaxType.RETURN, "RETURN", 2, 6),
            };

            AssertEqual(obs, expected);

            obs = GetObservedTokens("procedure a(b, c, d)\n\nreturn");
            expected = [
                    new(SyntaxType.PROCEDURE, "procedure", 1, 9),
                new(SyntaxType.NAME, "a", 1, 11),
                new(SyntaxType.LEFT_PAREN, "(", 1, 12),
                new(SyntaxType.NAME, "b", 1, 13),
                new(SyntaxType.COMMA, ",", 1, 14),
                new(SyntaxType.NAME, "c", 1, 16),
                new(SyntaxType.COMMA, ",", 1, 17),
                new(SyntaxType.NAME, "d", 1, 19),
                new(SyntaxType.RIGHT_PAREN, ")", 1, 20),
                new(SyntaxType.RETURN, "return", 3, 6),
            ];

            AssertEqual(obs, expected);
        }

        [TestMethod]
        public void TestIfElse()
        {
            var obs = GetObservedTokens(
                "if a <= b\n    b()\nelseif c\n    d()\nelse\n    e()\nendif"
            );
            var expected = new List<Token> {
                new(SyntaxType.IF, "if", 1, 2),
                new(SyntaxType.NAME, "a", 1, 4),
                new(SyntaxType.LE, "<=", 1, 7),
                new(SyntaxType.NAME, "b", 1, 9),
                new(SyntaxType.NAME, "b", 2, 5),
                new(SyntaxType.LEFT_PAREN, "(", 2, 6),
                new(SyntaxType.RIGHT_PAREN, ")", 2, 7),
                new(SyntaxType.ELSEIF, "elseif", 3, 6),
                new(SyntaxType.NAME, "c", 3, 8),
                new(SyntaxType.NAME, "d", 4, 5),
                new(SyntaxType.LEFT_PAREN, "(", 4, 6),
                new(SyntaxType.RIGHT_PAREN, ")", 4, 7),
                new(SyntaxType.ELSE, "else", 5, 4),
                new(SyntaxType.NAME, "e", 6, 5),
                new(SyntaxType.LEFT_PAREN, "(", 6, 6),
                new(SyntaxType.RIGHT_PAREN, ")", 6, 7),
                new(SyntaxType.ENDIF, "endif", 7, 5),
            };

            AssertEqual(obs, expected);
        }

        [TestMethod]
        public void TestConditional()
        {
            var obs = GetObservedTokens("iif(a, b, c)");
            var expected = new List<Token> {
                new(SyntaxType.IIF, "iif", 1, 3),
                new(SyntaxType.LEFT_PAREN, "(", 1, 4),
                new(SyntaxType.NAME, "a", 1, 5),
                new(SyntaxType.COMMA, ",", 1, 6),
                new(SyntaxType.NAME, "b", 1, 8),
                new(SyntaxType.COMMA, ",", 1, 9),
                new(SyntaxType.NAME, "c", 1, 11),
                new(SyntaxType.RIGHT_PAREN, ")", 1, 12),
            };

            AssertEqual(obs, expected);
        }


        private static List<Token> GetObservedTokens(string source)
        {
            lexer = new Lexer(source);

            var obs = new List<Token>();

            foreach (var token in lexer.GetTokens())
            {
                if (token.Type == SyntaxType.EOF)
                    break;

                obs.Add(token);
            }

            return obs;
        }

        private static bool AssertEqual(List<Token> obs, List<Token> expected)
        {
            if (obs.Count != expected.Count)
            {
                Assert.Fail($"Token count mismatch. Expected {expected.Count}, but got {obs.Count}.");
                return false;
            }
            for (int i = 0; i < obs.Count; i++)
            {
                if (obs[i].Type != expected[i].Type ||
                    obs[i].Text != expected[i].Text ||
                    obs[i].Line != expected[i].Line ||
                    obs[i].Start != expected[i].Start ||
                    obs[i].End != expected[i].End)
                {
                    Assert.Fail(
                        $"Token lines mismatch at index {i}. "
                        + $"Expected Token({expected[i].Type}, '{expected[i].Text}', {expected[i].Line}, {expected[i].Start}, {expected[i].End}), "
                        + $"but got Token({obs[i].Type}, '{obs[i].Text}', {obs[i].Line}, {obs[i].Start}, {obs[i].End})."
                    );
                    return false;
                }
            }
            return true;
        }
    }
}
